### PR TITLE
Add persisted and new_records methods to ActiveRecord::Associations::CollectionProxy.

### PR DESCRIPTION
**Pull request by stevepolitodesign**
This is a work in progress, and I wanted to get some feedback before I go too far.

This aims to extend the [methods added by has_many](https://guides.rubyonrails.org/association_basics.html#methods-added-by-has-many) and account for cases where a record is not yet saved but is still added to a collection.

## Usage

```ruby
class Post < ApplicationRecord
  has_many :comments
end

Post.first.comments.persisted
```

### Before

This will render all saved comments as well as the newly initialized comment.

```ruby
class PostsController < ApplicationController
  def show
    @post = Post.find(params[:id])
    @comment = @post.comments.build
    @comments = @post.comments
  end
end
```

```html+erb
<%= render @comments %>
```

### After

This will only render comments that have been saved to the database.

```ruby
class PostsController < ApplicationController
  def show
    @post = Post.find(params[:id])
    @comment = @post.comments.build
    @comments = @post.comments.persisted
  end
end
```

```html+erb
<%= render @comments %>
```

## To Do

- [x] Add and test `new_records` method.
- [x] Update changelong
- [x] Update the Guides
- [x] Update the API

### CODE REVIEW

File: activerecord/lib/active_record/associations/collection_proxy.rb [hunk 1]
```diff
@@ -1091,6 +1091,11 @@ def reset
         reset_scope
       end
 
+      def persisted
+        primary_key_column = @association.reflection.active_record_primary_key
+        load_target.select { |element| element[primary_key_column] != nil }
```
Comment (excid3) [inline activerecord/lib/active_record/associations/collection_proxy.rb:1096]: Would you be able to use the `persisted?` method instead?

```ruby
def persisted
  load_target.select(&:persisted?)
end
```

**Comment by DRBragg** [comment]
I dig this!

**Comment by excid3** [comment]
This is an area of Rails the catches a lot of people off guard, so I think this would be a useful method.

**Comment by ghiculescu** [comment]
I like this, did you consider the inverse (`comments.new_records`) too?

Please just remember to squash + add a changelog entry when you're happy with this.

**Comment by stevepolitodesign** [comment]
@ghiculescu I went ahead and addressed the following:

- Added a `new_records` method
- Updated the `CHANGELOG.md`
- Documented the methods in `activerecord/lib/active_record/associations/collection_proxy.rb`
- Squashed my commits.

**Questions**

Should I also update the guides since this would affect [4.3.1 Methods Added by has_many](https://guides.rubyonrails.org/association_basics.html#methods-added-by-has-many) and [4.4.1 Methods Added by has_and_belongs_to_many](https://guides.rubyonrails.org/association_basics.html#methods-added-by-has-and-belongs-to-many)

**Comment by ghiculescu** [comment]
> Should I also update the guides since this would affect [4.3.1 Methods Added by has_many](https://guides.rubyonrails.org/association_basics.html#methods-added-by-has-many) and [4.4.1 Methods Added by has_and_belongs_to_many](https://guides.rubyonrails.org/association_basics.html#methods-added-by-has-and-belongs-to-many)

Yeah I think so.

**Comment by stevepolitodesign** [comment]
Just a heads up, I'm currently unable to squash these into one commit due to some merge conflicts. I wouldn't want that to prevent this from being merged in, but I also don't know enough about git to resolve this at the moment.

**Comment by stevepolitodesign** [comment]
@rafaelfranca thank you for taking the time to review this. 

> ... I'm still uneasy to add those methods that looks like scopes and we would think they filter records in the database, but in reality they are filtering the records in memory.

Would it be better to create `persisted_collection` and/or `new_collection` methods? 

```ruby
book.persisted_authors
book.new_authors
```

I'm not sure if the `new_collection` method would have much value, but I think having a way to deliberately return persisted records is still valuable.

If you still have reservations about this proposal, maybe we should consider updating the guides to make developers aware that calling `collection` will return both persisted and newly instantiated records as I don't think this is made clear.

**Comment by ghiculescu** [comment]
Thinking out loud, could the new methods behave differently based on `loaded?`. For example, for `#persisted`:

```ruby
def persisted
  if loaded?
    load_target.select(&:persisted?)
  else
    # this isn't quite right, since you'd also want anything persisted `post.comments.create`, but something along these lines...
    all
  end
end
```

**Pull request closed without merging**